[Bug] Api Server长时间大批量推理后出现降速现象，最终出现CUDA runtime error: out of memory

### Checklist

- [X] 1. I have searched related issues but cannot get the expected help.
- [X] 2. The bug has not been fixed in the latest version.

### Describe the bug

通过最新的docker镜像启动一个api server，客户端使用api进行大批量、长时间的推理调用（总量20万次推理左右）。在经过大约10万次推理以后可观察到推理速度由最初的平均7it/s降至1it/s左右。
此时通过nvidia-smi观察显存占用，发现与调用开始时并无明显变化。
继续推理一段时间后，api server会报CUDA runtime error: out of memory /opt/lmdeploy/src/turbomind/utils/allocator.h:231 错误，然后可观察到服务重启，推理速度恢复正常。
推理速度降至1it/s时的显存状态：
![屏幕截图 2024-02-27 105210](https://github.com/InternLM/lmdeploy/assets/3341682/3126e5be-5b03-4155-87b2-f021a03255c6)


### Reproduction

参考 #636 实现了batch调用chat_completions_v1接口的方法，客户端通过该方法调用大批量的推理任务：
```python
import asyncio
from tqdm.asyncio import tqdm
from typing import Union, List, Dict, Optional
from dataclasses import dataclass, field
from lmdeploy.serve.openai.api_client import APIClient
from loguru import logger

@dataclass
class SingleChatReq:
    messages: Union[str, List[Dict[str, str]]]
    temperature: Optional[float] = 0.7
    top_p: Optional[float] = 1.0
    n: Optional[int] = 1
    max_tokens: Optional[int] = 512
    stop: Optional[Union[str, List[str]]] = None
    # stream: Optional[bool] = False  ## not supported yet
    presence_penalty: Optional[float] = 0.0
    frequency_penalty: Optional[float] = 0.0
    # user: Optional[str] = None  ## useless
    repetition_penalty: Optional[float] = 1.0
    # session_id: Optional[int] = -1  ## not supported yet
    ignore_eos: Optional[bool] = False
    skip_special_tokens: Optional[bool] = True
    extra_kwargs: Optional[dict] = field(default_factory=dict)


class ExtendBatchAPIClient(APIClient):
    def chat_completions_v1_batch(self,
                                  model: str,
                                  chat_requests: List[SingleChatReq],
                                  max_retries: int = 5) -> list:
        assert max_retries >= 0
        batch_size = len(chat_requests)
        batch_arr = chat_requests
        outputs = [""] * batch_size

        async def _try_req(single_req: SingleChatReq):
            try_count = 0
            while try_count <= max_retries:
                try:
                    response = await asyncio.to_thread(
                        lambda: next(
                            self.chat_completions_v1(
                                model=model,
                                messages=single_req.messages,
                                temperature=single_req.temperature,
                                top_p=single_req.top_p,
                                n=single_req.n,
                                max_tokens=single_req.max_tokens,
                                stop=single_req.stop,
                                presence_penalty=single_req.presence_penalty,
                                frequency_penalty=single_req.frequency_penalty,
                                repetition_penalty=single_req.repetition_penalty,
                                ignore_eos=single_req.ignore_eos,
                                skip_special_tokens=single_req.skip_special_tokens,
                                **single_req.extra_kwargs
                            )
                        )
                    )
                    return response
                except Exception as e:
                    try_count += 1
                    logger.warning(f"Request error, retry count: {try_count}. Exception: {e}")
                    if try_count <= max_retries:
                        await asyncio.sleep(1)
            raise RuntimeError("Max retries exceeded")

        async def get_response(_id):
            single_req = batch_arr[_id]
            response = await _try_req(single_req)
            outputs[_id] = response

        async def gather():
            tasks = [asyncio.ensure_future(get_response(_id)) for _id in range(batch_size)]
            try:
                await tqdm.gather(*tasks)
            except Exception as e:
                for t in tasks:
                    t.cancel()
                raise e

        asyncio.get_event_loop().run_until_complete(gather())
        return outputs
```

### Environment

```Shell
sys.platform: linux
Python: 3.8.10 (default, Nov 22 2023, 10:22:35) [GCC 9.4.0]
CUDA available: True
MUSA available: False
numpy_random_seed: 2147483648
GPU 0: NVIDIA GeForce RTX 4090 D
CUDA_HOME: /usr/local/cuda
NVCC: Cuda compilation tools, release 11.8, V11.8.89
GCC: x86_64-linux-gnu-gcc (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
PyTorch: 2.1.0+cu118
PyTorch compiling details: PyTorch built with:
  - GCC 9.3
  - C++ Version: 201703
  - Intel(R) oneAPI Math Kernel Library Version 2022.2-Product Build 20220804 for Intel(R) 64 architecture applications
  - Intel(R) MKL-DNN v3.1.1 (Git Hash 64f6bcbcbab628e96f33a62c3e975f8535a7bde4)
  - OpenMP 201511 (a.k.a. OpenMP 4.5)
  - LAPACK is enabled (usually provided by MKL)
  - NNPACK is enabled
  - CPU capability usage: AVX2
  - CUDA Runtime 11.8
  - NVCC architecture flags: -gencode;arch=compute_50,code=sm_50;-gencode;arch=compute_60,code=sm_60;-gencode;arch=compute_70,code=sm_70;-gencode;arch=compute_75,code=sm_75;-gencode;arch=compute_80,code=sm_80;-gencode;arch=compute_86,code=sm_86;-gencode;arch=compute_37,code=sm_37;-gencode;arch=compute_90,code=sm_90
  - CuDNN 8.7
  - Magma 2.6.1
  - Build settings: BLAS_INFO=mkl, BUILD_TYPE=Release, CUDA_VERSION=11.8, CUDNN_VERSION=8.7.0, CXX_COMPILER=/opt/rh/devtoolset-9/root/usr/bin/c++, CXX_FLAGS= -D_GLIBCXX_USE_CXX11_ABI=0 -fabi-version=11 -fvisibility-inlines-hidden -DUSE_PTHREADPOOL -DNDEBUG -DUSE_KINETO -DLIBKINETO_NOROCTRACER -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -DUSE_XNNPACK -DSYMBOLICATE_MOBILE_DEBUG_HANDLE -O2 -fPIC -Wall -Wextra -Werror=return-type -Werror=non-virtual-dtor -Werror=bool-operation -Wnarrowing -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-unused-parameter -Wno-unused-function -Wno-unused-result -Wno-strict-overflow -Wno-strict-aliasing -Wno-stringop-overflow -Wno-psabi -Wno-error=pedantic -Wno-error=old-style-cast -Wno-invalid-partial-specialization -Wno-unused-private-field -Wno-aligned-allocation-unavailable -Wno-missing-braces -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Werror=format -Werror=cast-function-type -Wno-stringop-overflow, LAPACK_INFO=mkl, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1, TORCH_DISABLE_GPU_ASSERTS=ON, TORCH_VERSION=2.1.0, USE_CUDA=ON, USE_CUDNN=ON, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=ON, USE_MPI=OFF, USE_NCCL=1, USE_NNPACK=ON, USE_OPENMP=ON, USE_ROCM=OFF,

TorchVision: 0.16.0+cu118
LMDeploy: 0.2.4+24ea5dc
transformers: 4.37.1
gradio: 3.50.2
fastapi: 0.109.2
pydantic: 2.6.1
```


### Error traceback

```Shell
服务端报错时的日志：

...
INFO:     192.168.16.15:55098 - "POST /v1/chat/completions HTTP/1.1" 200 OK
INFO:     192.168.16.15:55140 - "POST /v1/chat/completions HTTP/1.1" 200 OK
INFO:     192.168.16.15:55062 - "POST /v1/chat/completions HTTP/1.1" 200 OK
INFO:     192.168.16.15:55152 - "POST /v1/chat/completions HTTP/1.1" 200 OK
terminate called after throwing an instance of 'std::runtime_error'
  what():  [TM][ERROR] CUDA runtime error: out of memory /opt/lmdeploy/src/turbomind/utils/allocator.h:231 


=============================
== Triton Inference Server ==
=============================

NVIDIA Release 22.12 (build 50109463)
Triton Server Version 2.29.0

Copyright (c) 2018-2022, NVIDIA CORPORATION & AFFILIATES.  All rights reserved.

Various files include modifications (c) NVIDIA CORPORATION & AFFILIATES.  All rights reserved.

This container image and its contents are governed by the NVIDIA Deep Learning Container License.
By pulling and using the container, you accept the terms and conditions of this license:
https://developer.nvidia.com/ngc/nvidia-deep-learning-container-license

2024-02-26 10:24:00,949 - lmdeploy - WARNING - Best matched chat template name: internlm2-chat-7b
2024-02-26 10:24:00,968 - lmdeploy - WARNING - model_source: hf_model
2024-02-26 10:24:02,600 - lmdeploy - WARNING - model_config:

[llama]
model_name = internlm2-chat-7b
tensor_para_size = 1
head_num = 32
kv_head_num = 8
vocab_size = 92544
num_layer = 32
inter_size = 14336
norm_eps = 1e-05
attn_bias = 0
start_id = 1
end_id = 2
session_len = 32768
weight_type = fp16
rotary_embedding = 128
rope_theta = 1000000.0
size_per_head = 128
group_size = 0
max_batch_size = 128
max_context_token_num = 1
step_length = 1
cache_max_entry_count = 0.2
cache_block_seq_len = 128
cache_chunk_size = -1
num_tokens_per_iter = 0
max_prefill_iters = 1
extra_tokens_per_iter = 0
use_context_fmha = 1
quant_policy = 0
max_position_embeddings = 32768
rope_scaling_factor = 0.0
use_logn_attn = 0


2024-02-26 10:24:02,760 - lmdeploy - WARNING - get 259 model params
/opt/py38/lib/python3.8/site-packages/fuzzywuzzy/fuzz.py:11: UserWarning: Using slow pure-python SequenceMatcher. Install python-Levenshtein to remove this warning
  warnings.warn('Using slow pure-python SequenceMatcher. Install python-Levenshtein to remove this warning')
```
```

--- a/lmdeploy/serve/openai/api_server.py
+++ b/lmdeploy/serve/openai/api_server.py
@@ -1,7 +1,6 @@
 # Copyright (c) OpenMMLab. All rights reserved.
 import asyncio
 import os
-import random
 import time
 from http import HTTPStatus
 from typing import AsyncGenerator, List, Literal, Optional, Union
@@ -31,6 +30,7 @@ from lmdeploy.serve.qos_engine.qos_engine import QosEngine
 class VariableInterface:
     """A IO interface maintaining variables."""
     async_engine: AsyncEngine = None
+    session_id: int = 0
     api_keys: Optional[List[str]] = None
     qos_engine: QosEngine = None
     request_hosts = []
@@ -143,7 +143,8 @@ async def chat_completions_v1_qos(request: ChatCompletionRequestQos,
     - frequency_penalty (replaced with repetition_penalty)
     """
     if request.session_id == -1:
-        request.session_id = random.randint(1, 10086)
+        VariableInterface.session_id += 1
+        request.session_id = VariableInterface.session_id
     error_check_ret = await check_request(request)
     if error_check_ret is not None:
         return error_check_ret
@@ -291,7 +292,8 @@ async def chat_completions_v1(request: ChatCompletionRequest,
     - frequency_penalty (replaced with repetition_penalty)
     """
     if request.session_id == -1:
-        request.session_id = random.randint(1, 10086)
+        VariableInterface.session_id += 1
+        request.session_id = VariableInterface.session_id
     error_check_ret = await check_request(request)
     if error_check_ret is not None:
         return error_check_ret
@@ -449,7 +451,8 @@ async def completions_v1_qos(request: CompletionRequestQos,
     - frequency_penalty (replaced with repetition_penalty)
     """
     if request.session_id == -1:
-        request.session_id = random.randint(1, 10086)
+        VariableInterface.session_id += 1
+        request.session_id = VariableInterface.session_id
     error_check_ret = await check_request(request)
     if error_check_ret is not None:
         return error_check_ret
@@ -600,7 +603,8 @@ async def completions_v1(request: CompletionRequest,
     - frequency_penalty (replaced with repetition_penalty)
     """
     if request.session_id == -1:
-        request.session_id = random.randint(1, 10086)
+        VariableInterface.session_id += 1
+        request.session_id = VariableInterface.session_id
     error_check_ret = await check_request(request)
     if error_check_ret is not None:
         return error_check_ret
@@ -797,7 +801,8 @@ async def chat_interactive_v1_qos(request: GenerateRequestQos,
     - user_id (str): for qos; if not specified, will set to "default"
     """
     if request.session_id == -1:
-        request.session_id = random.randint(10087, 23333)
+        VariableInterface.session_id += 1
+        request.session_id = VariableInterface.session_id
 
     if VariableInterface.qos_engine is None:
         return create_error_response(
@@ -872,7 +877,8 @@ async def chat_interactive_v1(request: GenerateRequest,
         VariableInterface.async_engine.stop_session(request.session_id)
         return {'text': '', 'tokens': 0, 'finish_reason': None}
     if request.session_id == -1:
-        request.session_id = random.randint(10087, 23333)
+        VariableInterface.session_id += 1
+        request.session_id = VariableInterface.session_id
 
     async_engine = VariableInterface.async_engine
     sequence_start = async_engine.id2step.get(str(request.session_id), 0) == 0

